From cccb0d1db5a8e6e4da4c82394845da75e179506a Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:39:34 -0400
Subject: [ethernet] i40e: don't store user requested mode until we've validated it

Message-id: <1409658042-22689-214-git-send-email-sassmann@redhat.com>
Patchwork-id: 90418
O-Subject: [RHEL7.1 PATCH 213/281] i40e: don't store user requested mode until we've validated it
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jacob Keller <jacob.e.keller@intel.com>

This patch prevents the SIOCGHWTSTAMP ioctl from possibly returning bad
data, by not permanently storing the setting into the private
structure until after we've finished validating that we can support it.

Change-ID: Ib59f9b4f73f451d5a2e76fb8efa5d4271b218433
Signed-off-by: Jacob Keller <jacob.e.keller@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit d19af2afe70c11c17552e3290560037a8812f467)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_ptp.c b/drivers/net/ethernet/intel/i40e/i40e_ptp.c
index e5f558c..f7dded9 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_ptp.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_ptp.c
@@ -549,17 +549,20 @@ static int i40e_ptp_set_timestamp_mode(struct i40e_pf *pf,
  **/
 int i40e_ptp_set_ts_config(struct i40e_pf *pf, struct ifreq *ifr)
 {
- struct hwtstamp_config *config = &pf->tstamp_config;
+ struct hwtstamp_config config;
  int err;
 
- if (copy_from_user(config, ifr->ifr_data, sizeof(*config)))
+ if (copy_from_user(&config, ifr->ifr_data, sizeof(config)))
   return -EFAULT;
 
- err = i40e_ptp_set_timestamp_mode(pf, config);
+ err = i40e_ptp_set_timestamp_mode(pf, &config);
  if (err)
   return err;
 
- return copy_to_user(ifr->ifr_data, config, sizeof(*config)) ?
+ /* save these settings for future reference */
+ pf->tstamp_config = config;
+
+ return copy_to_user(ifr->ifr_data, &config, sizeof(config)) ?
   -EFAULT : 0;
 }
 
-- 
1.7.1